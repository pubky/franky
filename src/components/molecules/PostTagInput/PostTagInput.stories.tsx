import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { PostTagInput } from './PostTagInput';

const meta: Meta<typeof PostTagInput> = {
  title: 'Molecules/Post/PostTagInput',
  component: PostTagInput,
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'dark',
      values: [
        { name: 'dark', value: '#05050A' },
        { name: 'light', value: '#F2F2F7' },
      ],
    },
    docs: {
      description: {
        component:
          'Input component for adding new tags. Features a dashed border, placeholder text, and optional emoji picker button.',
      },
    },
  },
  args: {
    placeholder: 'add tag',
  },
  argTypes: {
    onChange: { action: 'changed' },
    onSubmit: { action: 'submitted' },
    onEmojiClick: { action: 'emoji clicked' },
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Playground: Story = {
  render: (args) => <PostTagInput {...args} />,
};

// Empty state (hint)
export const Empty: Story = {
  args: {
    value: '',
    showEmojiPicker: false,
  },
  render: (args) => <PostTagInput {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Empty input showing placeholder text.',
      },
    },
  },
};

// With text (filled)
export const Filled: Story = {
  args: {
    value: 'bitcoin',
    showEmojiPicker: false,
  },
  render: (args) => <PostTagInput {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Input with text entered.',
      },
    },
  },
};

// With emoji picker
export const WithEmojiPicker: Story = {
  args: {
    value: '',
    showEmojiPicker: true,
  },
  render: (args) => <PostTagInput {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Input with emoji picker button.',
      },
    },
  },
};

export const WithEmojiPickerFilled: Story = {
  args: {
    value: 'bitcoin',
    showEmojiPicker: true,
  },
  render: (args) => <PostTagInput {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Input with text and emoji picker button.',
      },
    },
  },
};

// Interactive demo
export const Interactive: Story = {
  args: {
    showEmojiPicker: true,
    onChange: (value) => console.log('Changed:', value),
    onSubmit: (value) => console.log('Submitted:', value),
    onEmojiClick: () => console.log('Emoji picker clicked'),
  },
  render: (args) => <PostTagInput {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Interactive input. Type and press Enter to submit. Check console for events.',
      },
    },
  },
};

// Custom placeholder
export const CustomPlaceholder: Story = {
  args: {
    placeholder: 'enter tag name...',
    showEmojiPicker: false,
  },
  render: (args) => <PostTagInput {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Input with custom placeholder text.',
      },
    },
  },
};

// Auto focus
export const AutoFocus: Story = {
  args: {
    autoFocus: true,
    showEmojiPicker: false,
  },
  render: (args) => <PostTagInput {...args} />,
  parameters: {
    docs: {
      description: {
        story: 'Input with auto focus enabled.',
      },
    },
  },
};
