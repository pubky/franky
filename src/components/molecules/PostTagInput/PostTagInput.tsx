import * as React from 'react';
import * as Atoms from '@/atoms';
import * as Libs from '@/libs';

export interface PostTagInputProps {
  /** Input ref */
  ref?: React.RefObject<HTMLInputElement>;
  /** Current input value */
  value?: string;
  /** Placeholder text */
  placeholder?: string;
  /** Show emoji picker button */
  showEmojiPicker?: boolean;
  /** Callback when input value changes */
  onChange?: (value: string) => void;
  /** Callback when Enter is pressed */
  onSubmit?: (value: string) => void;
  /** Callback when input loses focus */
  onBlur?: () => void;
  /** Callback when emoji picker is clicked */
  onEmojiClick?: () => void;
  /** Additional className */
  className?: string;
  /** Auto focus on mount */
  autoFocus?: boolean;
}

export function PostTagInput({
  value = '',
  placeholder = 'add tag',
  showEmojiPicker = false,
  onChange,
  onSubmit,
  onBlur,
  onEmojiClick,
  className,
  autoFocus = false,
}: PostTagInputProps) {
  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter' && value.trim()) {
      e.preventDefault();
      onSubmit?.(value.trim());
    }
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    onChange?.(e.target.value);
  };

  const InputBox = (
    <div
      className={Libs.cn(
        'relative flex h-8 items-center rounded-lg px-3',
        'border border-dashed border-input focus-within:border-white/80',
        className,
      )}
    >
      <Atoms.Input
        type="text"
        value={value}
        onChange={handleChange}
        onKeyDown={handleKeyDown}
        onBlur={onBlur}
        placeholder={placeholder}
        autoFocus={autoFocus}
        aria-label="Add new tag"
        className={Libs.cn(
          'flex-1 bg-transparent text-sm font-bold leading-5 outline-none caret-white',
          'border-none shadow-none ring-0 focus:ring-0 focus:outline-none focus:ring-offset-0 hover:outline-none',
          'p-0 h-6',
          showEmojiPicker ? 'pr-6' : undefined,
          'placeholder:text-input placeholder:font-bold',
          value ? 'text-secondary-foreground' : 'text-input',
        )}
      />

      {/* Emoji inside dashed area (both sizes) */}
      {showEmojiPicker && (
        <div className="pointer-events-auto absolute right-1 top-1/2 -translate-y-1/2">
          <Atoms.Button
            variant="secondary"
            size="icon"
            onClick={onEmojiClick}
            className="size-5 border-none p-1"
            style={{ boxShadow: '0px 1px 2px 0px rgba(5, 5, 10, 0.2)' }}
            aria-label="Open emoji picker"
          >
            <Libs.Smile className="size-3" strokeWidth={2} />
          </Atoms.Button>
        </div>
      )}
    </div>
  );
  return InputBox;
}
