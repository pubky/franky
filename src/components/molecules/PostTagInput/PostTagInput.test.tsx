import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { PostTagInput } from './PostTagInput';

// Mock @/libs with partial mock
vi.mock('@/libs', async (importOriginal) => {
  const actual = await importOriginal<typeof import('@/libs')>();
  return {
    ...actual,
    Smile: ({ className, strokeWidth }: { className?: string; strokeWidth?: number }) => (
      <div data-testid="smile-icon" className={className} data-stroke-width={strokeWidth}>
        Smile
      </div>
    ),
  };
});

describe('PostTagInput', () => {
  it('calls onChange when input value changes', () => {
    const mockOnChange = vi.fn();
    render(<PostTagInput onChange={mockOnChange} />);

    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 'bitcoin' } });

    expect(mockOnChange).toHaveBeenCalledWith('bitcoin');
  });

  it('calls onSubmit when Enter is pressed with value', () => {
    const mockOnSubmit = vi.fn();
    render(<PostTagInput value="bitcoin" onSubmit={mockOnSubmit} />);

    const input = screen.getByRole('textbox');
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(mockOnSubmit).toHaveBeenCalledWith('bitcoin');
  });

  it('does not call onSubmit when Enter is pressed with empty value', () => {
    const mockOnSubmit = vi.fn();
    render(<PostTagInput value="" onSubmit={mockOnSubmit} />);

    const input = screen.getByRole('textbox');
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(mockOnSubmit).not.toHaveBeenCalled();
  });

  it('trims whitespace before submitting', () => {
    const mockOnSubmit = vi.fn();
    render(<PostTagInput value="  bitcoin  " onSubmit={mockOnSubmit} />);

    const input = screen.getByRole('textbox');
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(mockOnSubmit).toHaveBeenCalledWith('bitcoin');
  });

  it('calls onEmojiClick when emoji button is clicked', () => {
    const mockOnEmojiClick = vi.fn();
    render(<PostTagInput showEmojiPicker onEmojiClick={mockOnEmojiClick} />);

    const emojiButton = screen.getByLabelText(/open emoji picker/i);
    fireEvent.click(emojiButton);

    expect(mockOnEmojiClick).toHaveBeenCalledTimes(1);
  });

  it('does not render emoji button when showEmojiPicker is false', () => {
    render(<PostTagInput showEmojiPicker={false} />);

    const emojiButton = screen.queryByLabelText(/open emoji picker/i);
    expect(emojiButton).not.toBeInTheDocument();
  });

  it('displays custom placeholder', () => {
    render(<PostTagInput placeholder="custom placeholder" />);

    const input = screen.getByPlaceholderText('custom placeholder');
    expect(input).toBeInTheDocument();
  });
});

describe('PostTagInput - Snapshots', () => {
  it('matches snapshot with empty state', () => {
    const { container } = render(<PostTagInput />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with filled state', () => {
    const { container } = render(<PostTagInput value="bitcoin" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with emoji picker', () => {
    const { container } = render(<PostTagInput showEmojiPicker />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with emoji picker and value', () => {
    const { container } = render(<PostTagInput value="bitcoin" showEmojiPicker />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
