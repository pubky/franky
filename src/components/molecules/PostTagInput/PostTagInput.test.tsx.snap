// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostTagInput - Snapshots > matches snapshot with emoji picker 1`] = `
<div
  class="relative flex h-8 items-center rounded-lg px-3 border border-dashed border-input focus-within:border-white/80"
>
  <input
    aria-label="Add new tag"
    class="file:text-foreground selection:bg-primary selection:text-primary-foreground border-input flex w-full min-w-0 rounded-md border transition-[color,box-shadow] file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:ring-0 flex-1 bg-transparent text-sm font-bold leading-5 outline-none caret-white border-none shadow-none ring-0 focus:ring-0 focus:outline-none focus:ring-offset-0 hover:outline-none p-0 h-6 pr-6 placeholder:text-input placeholder:font-bold text-input"
    data-testid="input"
    placeholder="add tag"
    type="text"
    value=""
  />
  <div
    class="pointer-events-auto absolute right-1 top-1/2 -translate-y-1/2"
  >
    <button
      aria-label="Open emoji picker"
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 size-5 border-none p-1"
      data-size="icon"
      data-slot="button"
      data-variant="secondary"
      style="box-shadow: 0px 1px 2px 0px rgba(5, 5, 10, 0.2);"
    >
      <div
        class="size-3"
        data-stroke-width="2"
        data-testid="smile-icon"
      >
        Smile
      </div>
    </button>
  </div>
</div>
`;

exports[`PostTagInput - Snapshots > matches snapshot with emoji picker and value 1`] = `
<div
  class="relative flex h-8 items-center rounded-lg px-3 border border-dashed border-input focus-within:border-white/80"
>
  <input
    aria-label="Add new tag"
    class="file:text-foreground selection:bg-primary selection:text-primary-foreground border-input flex w-full min-w-0 rounded-md border transition-[color,box-shadow] file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:ring-0 flex-1 bg-transparent text-sm font-bold leading-5 outline-none caret-white border-none shadow-none ring-0 focus:ring-0 focus:outline-none focus:ring-offset-0 hover:outline-none p-0 h-6 pr-6 placeholder:text-input placeholder:font-bold text-secondary-foreground"
    data-testid="input"
    placeholder="add tag"
    type="text"
    value="bitcoin"
  />
  <div
    class="pointer-events-auto absolute right-1 top-1/2 -translate-y-1/2"
  >
    <button
      aria-label="Open emoji picker"
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 size-5 border-none p-1"
      data-size="icon"
      data-slot="button"
      data-variant="secondary"
      style="box-shadow: 0px 1px 2px 0px rgba(5, 5, 10, 0.2);"
    >
      <div
        class="size-3"
        data-stroke-width="2"
        data-testid="smile-icon"
      >
        Smile
      </div>
    </button>
  </div>
</div>
`;

exports[`PostTagInput - Snapshots > matches snapshot with empty state 1`] = `
<div
  class="relative flex h-8 items-center rounded-lg px-3 border border-dashed border-input focus-within:border-white/80"
>
  <input
    aria-label="Add new tag"
    class="file:text-foreground selection:bg-primary selection:text-primary-foreground border-input flex w-full min-w-0 rounded-md border transition-[color,box-shadow] file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:ring-0 flex-1 bg-transparent text-sm font-bold leading-5 outline-none caret-white border-none shadow-none ring-0 focus:ring-0 focus:outline-none focus:ring-offset-0 hover:outline-none p-0 h-6 placeholder:text-input placeholder:font-bold text-input"
    data-testid="input"
    placeholder="add tag"
    type="text"
    value=""
  />
</div>
`;

exports[`PostTagInput - Snapshots > matches snapshot with filled state 1`] = `
<div
  class="relative flex h-8 items-center rounded-lg px-3 border border-dashed border-input focus-within:border-white/80"
>
  <input
    aria-label="Add new tag"
    class="file:text-foreground selection:bg-primary selection:text-primary-foreground border-input flex w-full min-w-0 rounded-md border transition-[color,box-shadow] file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:ring-0 flex-1 bg-transparent text-sm font-bold leading-5 outline-none caret-white border-none shadow-none ring-0 focus:ring-0 focus:outline-none focus:ring-offset-0 hover:outline-none p-0 h-6 placeholder:text-input placeholder:font-bold text-secondary-foreground"
    data-testid="input"
    placeholder="add tag"
    type="text"
    value="bitcoin"
  />
</div>
`;
